mongo/single_result.go

```diff
@@ -74,32 +78,37 @@ func (sr *SingleResult) Decode(v interface{}) error {
 		return sr.err
 	}
 
-	dec := getDecoder(sr.rdr, sr.bsonOpts, sr.reg)
+	if sr.rdr.Type == bson.TypeEmbeddedDocument {
+		dec := getDecoder(sr.rdr.Value, sr.bsonOpts, sr.reg)
 
-	return dec.Decode(v)
+		return dec.Decode(v)
+	}
+
+	return sr.rdr.Unmarshal(v)
```

Comment:
We need to use the configured `Registry` and `BSONOptions` when unmarshaling the `RawValue` here.

Reply:
It makes sense to decode distinct data with a registry, but the BSONOptions are specific to `bson.Decoder` which expects the value to decode into to be a map / struct. E.g., the following would throw a runtime error:

```go
var i32s []int32
distincResult.Decode(i32s)
```

Reply:
You're correct, there's currently no way to do that. However, when combined with the changes in https://github.com/mongodb/mongo-go-driver/pull/1584, the following code using a `bson.Decoder` to decode a value works:
```go
func (dr *DistinctResult) Decode(v any) error {
	doc := bsoncore.NewDocumentBuilder().
		AppendValue("Arr", bsoncore.Value{
			Type: bsoncore.TypeArray,
			Data: dr.arr,
		}).Build()

	dec := getDecoder(doc, dr.opts, dr.reg)

	xs := struct {
		Arr any
	}{
		Arr: v,
	}
	return dec.Decode(&xs)
}
```
I created [GODRIVER-3205](https://jira.mongodb.org/browse/GODRIVER-3205) and added it to the Go Driver 2.0 epic to track that improvement, so we can consider that is resolved for now.